### DISCUSSION TITLE
Readonly classes

### DISCUSSION THREAD

**Discussion by Sunlighter** [dotnet/csharplang#4696] [Language Ideas] [answered]
I'd like to be able to define a `readonly` class or struct, e.g., `public readonly class MyClass { ... }`

This would be the same as a regular class except that all the fields would have to be `readonly` and their types would have to be either primitive types which are known to be immutable, or other `readonly` classes or structs. Or delegates. Failure to meet this requirement would cause a compiler error.

The Immutable Collections library would have to be modified so that `ImmutableList<T>` and the like are all `readonly` classes.

The benefit would be that it is easy to guarantee that objects of a given class are immutable.

An attribute would probably have to be defined in `System.Runtime.CompilerServices` so that a class could be marked as `readonly` in the metadata by having that attribute.

There would probably also need to be a generic constraint, so that you could write `public readonly class MyClass<T> where T : readonly { ... }`

**Comment by jnm2** [comment, marked as answer]
Me too! It's proposed here and is on @333fred's radar: https://github.com/dotnet/csharplang/issues/3885

I only want it to mean the same thing it means on `readonly struct` (C# 7.2+), which is to say, shallow immutability not deep immutability. `immutable` would be a better keyword for the latter IMO. For deep immutability, which would also be neat, see  https://github.com/dotnet/csharplang/issues/2543 and https://github.com/dotnet/csharplang/discussions/4015.

**Reply by Sunlighter** [reply]
Yeah, I was thinking `readonly` should be deep. I did not know about the previous proposals.

However, deep `readonly` might not be possible. For one thing, existing code might keep immutable collections of mutable objects. (Or objects might be `readonly` but not marked as such.) If `ImmutableList<T>` were to be made deeply readonly by having the `where T : readonly` constraint on it, such code would be broken. But if it doesn&rsquo;t have that constraint, `readonly` is shallow for classes that use `ImmutableList<T>` (and all immutable collections).

This could be fixed with a new &ldquo;transitive&rdquo; constraint syntax like `public class MyClass<T> readonly where T : readonly { ... }` but that seems complicated. Or you could have two `ImmutableList<T>` classes, but one wonders about naming. `ReallyImmutableList<T>`?

There are already problems like that with `IComparable<T>`. You can&rsquo;t define a `MyClass<T>` which may or may not implement `IComparable<MyClass<T>>` depending on whether `T` implements `IComparable<T>`. Maybe it would be nice if you could do that, but I don&rsquo;t know what the syntax would look like. It becomes like SFINAE in C++ templates. I suppose you could throw runtime exceptions.

**Reply by HaloFour** [reply]
> Yeah, I was thinking `readonly` should be deep.

I definitely think there'd be value in providing deep immutability in the language and runtime, but where the `readonly` modifier can be applied today only provides shallow immutability.  So, in my opinion, it'd be better to choose a different keyword for deep immutability.

**Comment by theunrepentantgeek** [comment]
Why is _deep immutability_ a requirement? 

I understand the need for _apparent_ immutability where the value represented doesn't appear to change (and can therefore be shared), but there are good reasons why sometimes an readonly type might have something mutable buried within. 

To illustrate, consider an immutable tree structure where calculation of a hash-code might be quite expensive. It's not always needed, so calculation in the constructor (or builder) just makes things perform poorly, but once calculated the result will never change and it's useful to cache the result using `Lazy<int>` for reuse.

It's also an advantage to pass off something mutable (say, a `List<T>`) into an readonly data structure where it's not going to be modified - if that required laboriously copying the items across into a different data structure, you're again killing performance for no actual gain.

**Reply by CyrusNajmabadi** [reply]
I agree with @theunrepentantgeek here :-)

**Reply by Sunlighter** [reply]
*Deep immutability* can be a nice guarantee to have in some cases.

One is encapsulation: if you have a (mutable) class with a private `List` and you want to have a property that returns that `List`, then theoretically you should make a copy, because otherwise whoever uses the property to get the list can modify it, and your class would not be able to maintain its invariant. But with `ImmutableList` that is not a problem.

Another case is implementing undo: you can keep previous versions of an immutable object, and the previous versions can share any sub-objects which have not changed. But if the object has mutable parts, then when you modify the current version, you are also modifying the previous versions, and that defeats the purpose of having &ldquo;undo.&rdquo;

In cases like that having only shallow readonly provides a false sense of security because some other part of the code could still make deep modifications that interfere with things. A deep readonly guarantees that won&rsquo;t happen.

It does make sense that we want to treat `Lazy<T>` as immutable even though it really isn&rsquo;t. Its value never changes, but it may not have calculated its value yet, that&rsquo;s all. That might require special handling.

C++ introduced `const`, then later had to add `mutable` to express the idea that something might change even though it is &ldquo;logically `const`.&rdquo; `Lazy<T>` is a good use case for that.

However, proposing a new `mutable` keyword for C# is difficult because it may break existing code. It might be better to create a `System.Mutable<T>` struct which has a mutable `T` but is marked `readonly` somehow and thus allows you to &ldquo;cheat&rdquo; in a controlled way.

(Maybe `System.Mutable<T>` should be a class because a `readonly struct` already exists and has certain features that would be broken by `System.Mutable<T>`.)

**Reply by Sunlighter** [reply]
It occurred to me later that one good use for deep immutability is for keys in sets and dictionaries. If you put a mutable object in a set, or use it as a key in a dictionary, and then change the object, it might no longer be in the correct place in the set or dictionary...

**Reply by DavidArno** [reply]
```cs
readonly List<string> _field;
```
To my way of thinking, that `readonly` for `_field` is pretty much meaningless as all it offers is an "assurance" that the reference won't change. The list itself is free to change in any way it wants. So even shallow read only classes would be useful here as I could then have an analyzer that warned against using `readonly` on fully mutable classes. 

And whilst there are cases where shallow immutability is useful, such as for handling lazy initialisation, there are also many use cases for true immutability in class as well as structs. Both features have their place and both would be welcome additions to the language.

And I completely disagree with "*if that required laboriously copying the items across into a different data structure, you're again killing performance for no actual gain*". There is zero point to a pseudo readonly wrapper over a `List<T>` that exposes the list as that list remains publicly mutable. The gain of eg only exposing it via an iterator property is massive: it guarantees it's publicly read only. That trumps performance every time in my book unless that performance hit is a bottleneck. And if it isn't a bottleneck, you've opted for premature optimisation and made the code harder to test and maintain.

**Comment by foxesknow** [comment]
I'm looking forward to the inevitable "write an analyzer to do it" answer...

**Reply by spydacarnage** [reply]
Well, if you insist...

You could write an analyzer to report on mutations occurring...

**Reply by DavidArno** [reply]
Read only classes would make it easier to write analyzers. Thankfully I don't think the "write an analyzer to do it" brigade are yet proposing we write analysers to make it easier to write more analyzers. We'd be into "turtle analyzers" territory there, [with analyzers all the way down](https://en.wikipedia.org/wiki/Turtles_all_the_way_down) 😄

**Comment by YairHalberstadt** [comment]
The problem with deep immutability is that it's almost impossible to retrofit onto an existing framework.

To highlight the problem consider an immutable class that contains a collection:

```csharp
immutable record MyRecord(IEnumerable<int> Collection);
```

I think it's pretty reasonable to want to write such code.

the problem is that implementations of `IEnumerable<T>` can be mutable, so you'd have to create a new interface that is exactly the same as `IEnumerable<T>` except it guarentees the implementor must be `immutable`:

```csharp
immutable interface IImmutableEnumerable<T> : IEnumerable<T> {}
immutable record MyRecord(IImmutableEnumerable<int> Collection);
```

Now I want to do an immutable update to the data in `MyRecord.Collection` so I use linq:

```csharp
var myNewRecord = myRecord with { Collection = myRecord.Collection.Select(x => x * x) };
```

But that won't work - `Select` returns an `IEnumerable<int>` not an `IImmutableEnumerable<int>`.

So now we need to reimplement LINQ in exactly the same way, but taking `IImmutableEnumerable` instead of `IEnumerable`.

```csharp
public static IImmutableEnumerable<TResult> Select<TSource, TResult>(this IImmutableEnumerable<TSource> source, Func<TSource, TResult> selector) => ...
```

Does that work?

No!

selector might not be a pure function, which would mean `myRecord.Collection` would return different results each time it was iterated - in other words it wouldn't be `immutable`. E.g. I could do `collection.Select(x => new Random().Next())`

So now we need to introduce `immutable delegates`:

```csharp
immutable delegate TResult ImmutableFunc<TSource, TResult(TSource source, TResult result);
public static IImmutableEnumerable<TResult> Select<TSource, TResult>(this IImmutableEnumerable<TSource> source, ImmutableFunc<TSource, TResult> selector) => ...
```

What can be assigned to an `immutable delegate`? Well a pure function of course.

What's a pure function?

Well it's one that doesn't assign any fields, and **only calls other pure functions**.

Which means that you need to annotate the entirety of the BCL with which methods are pure and which are not before it becomes practical to write pure functions.

Then you'll need to handle more complex conditional cases involving generics - "this method is pure, if and only if this method on the type parameter it calls is pure", "this type is immutable if and only if its type parameter is immutable".

So properly supporting immutability requires a huge number of concepts:

- immutable class, structs, interfaces, and delegates
- pure functions
- conditional purity and immutability

It also requires:

- redefining many/most interfaces with immutable versions
- reimplementing many/most APIs with immutable versions
- annotating most APIs with `pure`

You could avoid redefining interfaces and APIs by making a more complex conditional annotation system, where `Select` could say:
"If the implementation of `IEnumerable<TSource> and Func<TSource, TResult>` happen to be immutable, then the return value will also be immutable. This comes at the cost of an even greater concept count and complexity, and even more annotation work.

**Reply by Sunlighter** [reply]
I think it would be sufficient to keep the deep immutability confined to data types. If you want to write a deeply immutable type then its member variables must also be deeply immutable (or primitive types).

That would prevent an immutable record from having an `IEnumerable<T>` as a member, and I think that's fine. I'd probably just use an `ImmutableList<T>` instead and use the extension method `ToImmutableList()`.

Pure functions are a different can of worms because they can't modify *anything.* For me, that's overkill. It's sufficient if I can define a type in such a way that instances of that type cannot be modified.

I do see the need for conditional propagation, i.e., `ImmutableList<T>` is deeply immutable if and only if T is deeply immutable. I admit that's somewhat complex, but it at least prevents us from having to define two `ImmutableList` types (one shallow, one deep). And, as I mentioned, I might want a `System.Mutable<T>` class that pretended to be readonly but really wasn't. That could be used for implementing laziness, caching, and the like.

However, I don't want to turn C# into Haskell. I just want to be able to define immutable types.

**Reply by YairHalberstadt** [reply]
> I'd probably just use an ImmutableList<T>

Not being able to store any abstraction in an immutable type would be severely limiting, to the point at which I'm not sure the benefits are worth it.
